homeassistant/components/tplink/switch.py

```diff
@@ -32,131 +34,51 @@ async def async_setup_entry(
 
 
-class SmartPlugLedSwitch(CoordinatedTPLinkEntity, SwitchEntity):
-    """Representation of switch for the LED of a TPLink Smart Plug."""
-
-    _attr_translation_key = "led"
-    _attr_entity_category = EntityCategory.CONFIG
-
-    def __init__(
-        self, device: Device, coordinator: TPLinkDataUpdateCoordinator, led_module: Led
-    ) -> None:
-        """Initialize the LED switch."""
-        super().__init__(device, coordinator)
-        self._attr_unique_id = f"{device.mac}_led"
-        self._led_module = led_module
-        self._async_update_attrs()
-
-    @async_refresh_after
-    async def async_turn_on(self, **kwargs: Any) -> None:
-        """Turn the LED switch on."""
-        await self._led_module.set_led(True)
-
-    @async_refresh_after
-    async def async_turn_off(self, **kwargs: Any) -> None:
-        """Turn the LED switch off."""
-        await self._led_module.set_led(False)
-
-    @callback
-    def _async_update_attrs(self) -> None:
-        """Update the entity's attributes."""
-        self._attr_is_on = self._led_module.led
-
-    @callback
-    def _handle_coordinator_update(self) -> None:
-        """Handle updated data from the coordinator."""
-        self._async_update_attrs()
-        super()._handle_coordinator_update()
-
-
-class SmartPlugSwitch(CoordinatedTPLinkEntity, SwitchEntity):
-    """Representation of a TPLink Smart Plug switch."""
-
-    _attr_name: str | None = None
+class Switch(CoordinatedTPLinkEntity, SwitchEntity):
+    """Representation of a feature-based TPLink sensor."""
 
     def __init__(
         self,
         device: Device,
         coordinator: TPLinkDataUpdateCoordinator,
+        feature: Feature,
+        parent: Device | None = None,
     ) -> None:
         """Initialize the switch."""
-        super().__init__(device, coordinator)
-        # For backwards compat with pyHS100
-        self._attr_unique_id = legacy_device_id(device)
-        self._async_update_attrs()
+        super().__init__(device, coordinator, feature=feature, parent=parent)
+        self._feature: Feature
 
-    @async_refresh_after
-    async def async_turn_on(self, **kwargs: Any) -> None:
-        """Turn the switch on."""
-        await self.device.turn_on()
-
-    @async_refresh_after
-    async def async_turn_off(self, **kwargs: Any) -> None:
-        """Turn the switch off."""
-        await self.device.turn_off()
-
-    @callback
-    def _async_update_attrs(self) -> None:
-        """Update the entity's attributes."""
-        self._attr_is_on = self.device.is_on
+        # Use the device name for the primary switch control
+        if feature.category is Feature.Category.Primary:
+            self._attr_name = None
 
-    @callback
-    def _handle_coordinator_update(self) -> None:
-        """Handle updated data from the coordinator."""
+        self.entity_description = _description_for_feature(
+            SwitchEntityDescription, feature
+        )
         self._async_update_attrs()
```

Comment:
I thought it'd be fine to avoid calling this in `__init__`, but even adding `update_before_add=True` to `async_add_entities()` does not seem to be enough to get the entity updated for tests, so I suppose this needs to be called.

Or is it better to keep the `_handle_coordinator_update` to call this?

Reply:
As a side effect for calling the update here makes us not to catch potential exceptions like we are doing in a single call-site for updates, see https://github.com/python-kasa/python-kasa/pull/933

Reply:
> I thought it'd be fine to avoid calling this in `__init__`, but even adding `update_before_add=True` to `async_add_entities()` does not seem to be enough to get the entity updated for tests, so I suppose this needs to be called.
> 
> 
> 
> Or is it better to keep the `_handle_coordinator_update` to call this?

Update before add is highly undesirable because it slows down platform setup and makes everything much more complicated since it has to do another update after the first refresh

Reply:
Thanks for the explanation, I'll look into what's the best way to handle this nicely.